libraries/tock-cells/src/map_cell.rs

```diff
@@ -124,22 +263,33 @@ impl<T> MapCell<T> {
     /// // but potentially changed.
     /// assert_eq!(y.take(), Some(1235));
     /// ```
+    ///
+    /// # Panics
+    /// If debug assertions are enabled, this panics if the `MapCell`'s contents are already borrowed.
+    #[inline(always)]
     pub fn map<F, R>(&self, closure: F) -> Option<R>
     where
         F: FnOnce(&mut T) -> R,
     {
-        if self.is_some() {
-            self.occupied.set(false);
-            let valref = unsafe { &mut *self.val.get() };
-            // TODO: change to valref.get_mut() once stabilized [#53491](https://github.com/rust-lang/rust/issues/53491)
-            let res = closure(unsafe { &mut *valref.as_mut_ptr() });
-            self.occupied.set(true);
-            Some(res)
-        } else {
-            None
-        }
+        debug_assert_not_borrowed!(self);
+        (self.occupied.get() == MapCellState::Init).then(move || {
+            self.occupied.set(MapCellState::Borrowed);
+            // `occupied` is reset to initialized at the end of scope,
+            // even if a panic occurs in `closure`.
+            struct ResetToInit<'a>(&'a Cell<MapCellState>);
+            impl Drop for ResetToInit<'_> {
+                #[inline(always)]
+                fn drop(&mut self) {
+                    self.0.set(MapCellState::Init);
+                }
+            }
+            let _reset_to_init = ResetToInit(&self.occupied);
```

Comment:
While this unwind-safety is nice (especially for unit tests), the Tock kernel has generally been designed with the assumption that the kernel will be built with panic=abort, and I don't think we aim to guarantee sound operation if unwinding panics are used / caught. And this presumably does come at some size cost, so may be worth considering whether it makes sense to include. 

That said, I don't want this PR to get held up on a general discussion of whether we should provide unwind-safety in Tock.

Reply:
> And this presumably does come at some size cost

The size cost should only be there if it needs to catch an unwind to reset the state. This is only true if you have `-Cpanic=unwind` _and_ the compiler cannot prove the closure is panic-less. Otherwise it's the same as writing the normal `self.occupied.set(MapCellState::Init)` - note the `#[inline(always)]` to ensure this is the case. In other words, it won't affect Tock users that write `-Cpanic=abort` but will help users in unit tests and systems with `-Cpanic=unwind`.

When it comes to fundamental data structures like this, I think we should lean towards caution here for unwind-safety. Otherwise, it would cause missed drop glue and weird behavior as it would then forever think it is still borrowed - fortunately not unsoundness though.